extensions/amp-analytics/0.1/test/test-transport.js

```diff
@@ -48,69 +57,192 @@ describe('amp-analytics.transport', () => {
         'sendRequestUsingImage call count').to.equal(expectedImageCalls);
   }
 
+  function expectAllUnique(numArray) {
+    if (!numArray) {
+      return;
+    }
+    expect(numArray.length).to.equal(new Set(numArray).size);
+  }
+
+  it('prefers cross-domain iframe over beacon, xhrpost, and image', () => {
+    setupStubs(true, true, true, true);
+    transport.sendRequest(window, 'https://example.com/test', {
+      iframe: 'https://example.com/test',
+    });
+    assertCallCounts(1, 0, 0, 0);
+  });
+
   it('prefers beacon over xhrpost and image', () => {
-    setupStubs(true, true);
-    sendRequest(window, 'https://example.com/test', {
+    setupStubs(true, true, true, true);
+    transport.sendRequest(window, 'https://example.com/test', {
       beacon: true, xhrpost: true, image: true,
     });
-    assertCallCounts(1, 0, 0);
+    assertCallCounts(0, 1, 0, 0);
   });
 
   it('prefers xhrpost over image', () => {
-    setupStubs(true, true);
-    sendRequest(window, 'https://example.com/test', {
+    setupStubs(true, true, true, true);
+    transport.sendRequest(window, 'https://example.com/test', {
       beacon: false, xhrpost: true, image: true,
     });
-    assertCallCounts(0, 1, 0);
+    assertCallCounts(0, 0, 1, 0);
   });
 
   it('reluctantly uses image if nothing else is enabled', () => {
-    setupStubs(true, true);
-    sendRequest(window, 'https://example.com/test', {
+    setupStubs(true, true, true, true);
+    transport.sendRequest(window, 'https://example.com/test', {
       image: true,
     });
-    assertCallCounts(0, 0, 1);
+    assertCallCounts(0, 0, 0, 1);
   });
 
   it('falls back to xhrpost when enabled and beacon is not available', () => {
-    setupStubs(false, true);
-    sendRequest(window, 'https://example.com/test', {
+    setupStubs(false, false, false, true);
+    transport.sendRequest(window, 'https://example.com/test', {
       beacon: true, xhrpost: true, image: true,
     });
-    assertCallCounts(1, 1, 0);
+    assertCallCounts(0, 1, 1, 0);
   });
 
   it('falls back to image when beacon not found and xhr disabled', () => {
-    setupStubs(false, true);
-    sendRequest(window, 'https://example.com/test', {
+    setupStubs(false, false, false, true);
+    transport.sendRequest(window, 'https://example.com/test', {
       beacon: true, xhrpost: false, image: true,
     });
-    assertCallCounts(1, 0, 1);
+    assertCallCounts(0, 1, 0, 1);
   });
 
   it('falls back to image when beacon and xhr are not available', () => {
-    setupStubs(false, false);
-    sendRequest(window, 'https://example.com/test', {
+    setupStubs(false, false, false, false);
+    transport.sendRequest(window, 'https://example.com/test', {
       beacon: true, xhrpost: true, image: true,
     });
-    assertCallCounts(1, 1, 1);
+    assertCallCounts(0, 1, 1, 1);
+  });
+
+  it('must create xframe before sending message to it', () => {
+    expect(() => {
+      transport.sendRequest(window, 'https://example.com/test', {
```

Comment:
Copying conversation from old PR:

@keithwrightbos: test referencing window makes me nervous and implies your test may not be hermetic...

Me: I need to use _a_ window object, to hang iframes on. Also the existing tests need one, to do stuff like win.navigator.sendBeacon(request, '');
But I agree it would be best to somehow reset it, or use some type of "test window" object that lives in the sinon sandbox. Do you know a way to do this?